### PR TITLE
Set the scene for phase 1 of the new trait encoding.

### PR DESCRIPTION
**Pull request by adriaanm**
The next PR will introduce the Fields phase and the accompanying refactorings. A lot of these cleanups and refactorings are motivated/required by that.

### CODE REVIEW

**Comment by adriaanm** [comment]
I'll resuscitate your comments, @retronym.

**Comment by adriaanm** [comment]
Sneak preview for step 2: https://github.com/adriaanm/scala/compare/fields-prep...traits-late-fields?expand=1

File: src/reflect/scala/reflect/internal/Internals.scala [hunk 1]
```diff
@@ -60,19 +60,7 @@ trait Internals extends api.Internals {
...
-            case _: DefTree => change(tree.symbol.moduleClass)
-            case _          => // do nothing
-          }
-          super.traverse(tree)
-        }
-      }
-      changeOwnerAndModuleClassTraverser.traverse(tree)
-      tree
-    }
+    def changeOwner(tree: Tree, prev: Symbol, next: Symbol): tree.type = { new ChangeOwnerTraverser(prev, next).traverse(tree); tree }
```
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Internals.scala:63]: No longer need the override, as per @retronym's review comment -- we do the right thing in ChangeOwnerTraverser

**Comment by adriaanm** [comment]
Temporarily dropped the last commit for renovations. TODO items from @retronym's review:

 - [ ] Needs a test along the lines of:

```
scala> implicit def s2i(s: String): Int = s.length; trait Base { def foo: Int }; trait Sub extends Base { def foo = "" }
warning: there was one feature warning; re-run with -feature for details
s2i: (s: String)Int
defined trait Base
defined trait Sub

scala> implicit def s2i(s: String): Int = s.length; trait Base { def foo: Int }; trait Sub extends Base { val foo = "" }
<console>:9: error: overriding method foo in trait Base of type => Int;
 value foo has incompatible type
       implicit def s2i(s: String): Int = s.length; trait Base { def foo: Int }; trait Sub extends Base { val foo = "" }
```

I take it the second example will work after this change.

  - [ ] It might be worth spelling out in more detail that we're using the result type of the overrriden member as the expected type for the RHS of the val. We then use the type of the RHS (which may be more specific) as the result type of the overriding member.

  - [ ]This seems clever, but is different from the approach used when finding the overriden symbol of a DefDef:

```
def overriddenSymbol(resTp: Type) = {
        lazy val schema: Type = methodTypeSchema(resTp) // OPT create once. Must be lazy to avoid cycles in neg/t5093.scala
        intersectionType(methOwner.info.parents).nonPrivateMember(meth.name).filter { sym =>
          sym != NoSymbol && (site.memberType(sym) matches schema)
        }
      }
```

I'd like to unify the approach.

**Comment by adriaanm** [comment]
I've verified that bytecode is identical between quick & strap, when compiling with locker built from 2.12.x HEAD. I'm probably going to defer the val type inference commit to another PR, to avoid overloading this one.

**Comment by adriaanm** [comment]
ready for review by @retronym & @lrytz

**Comment by lrytz** [comment]
LGTM (!)

**Comment by adriaanm** [comment]
Thanks, @lrytz! Since this PR touches a lot of stuff, let's make merge conditional on LGTMs from both reviewers.

**Comment by retronym** [comment]
LGTM

**Comment by adriaanm** [comment]
Thanks for the quick turnaround, guys!

File: src/compiler/scala/tools/nsc/transform/LambdaLift.scala [hunk 2]
```diff
@@ -374,41 +377,45 @@ abstract class LambdaLift extends InfoTransform {
...
-            //   DefDef(getter, rhs) setPos tree.pos setType NoType
-            // }
-            // val newDefs = if (sym.isTrait) freeParams ::: (ps map paramGetter) else freeParams
-            deriveClassDef(tree)(impl => deriveTemplate(impl)(_ ::: freeParams))
-        }
-      case None =>
-        tree
-    }
 
 /*  SI-6231: Something like this will be necessary to eliminate the implementation
```
Comment (soc) [inline src/compiler/scala/tools/nsc/transform/LambdaLift.scala:420]: Is this comment obsolete? After the changes, `paramGetter` isn't even mentioned in this class anymore.

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/LambdaLift.scala:420]: Good catch! It's not obsolete yet, unfortunately. I deleted this comment a bit prematurely. I have a different [branch](https://github.com/adriaanm/scala/tree/traits-late-fields-lambdalift) with this fix in the works, and this change was left over from the preliminary work on that.

Comment (soc) [inline src/compiler/scala/tools/nsc/transform/LambdaLift.scala:420]: Alright, thanks for the details!

**Pull request merged**